src/node/stage/approved.rs

```diff
@@ -51,99 +49,183 @@ pub(crate) struct Approved {
     // them after we are demoted.
     relocate_promise: Option<Bytes>,
-    comm: Comm,
     msg_filter: MessageFilter,
-    timer: Timer,
 }
 
 impl Approved {
-    // Create the approved stage for a regular node.
+    // Creates the approved state for the first node in the network
+    pub fn first_node(node_info: NodeInfo, our_addr: SocketAddr) -> Result<Self> {
+        let peer = Peer::new(node_info.name(), our_addr);
+        let (shared_state, section_key_share) = SharedState::first_node(peer)?;
+
+        Ok(Self::new(shared_state, Some(section_key_share), node_info))
+    }
+
+    // Creates the approved state for a regular node.
     pub fn new(
-        comm: Comm,
         shared_state: SharedState,
         section_key_share: Option<SectionKeyShare>,
         node_info: NodeInfo,
-        timer: Timer,
-    ) -> Result<Self> {
+    ) -> Self {
         let section_keys_provider = SectionKeysProvider::new(section_key_share);
 
-        Ok(Self {
+        Self {
             node_info,
             shared_state,
             section_keys_provider,
             vote_accumulator: Default::default(),
             section_update_barrier: Default::default(),
             dkg_voter: Default::default(),
             relocate_promise: None,
-            comm,
             msg_filter: MessageFilter::new(),
-            timer,
-        })
+        }
     }
 
-    pub async fn process_message(
+    pub fn handle_message(
         &mut self,
-        sender: SocketAddr,
+        sender: Option<SocketAddr>,
         msg: Message,
-    ) -> Result<Option<Bootstrapping>> {
+    ) -> Result<Vec<Command>> {
+        let mut commands = vec![];
+
+        // Check if the message is for us.
+        let in_dst_location = msg
+            .dst()
+            .contains(&self.node_info.name(), self.shared_state.our_prefix());
+        if !in_dst_location || msg.dst().is_section() {
+            // Relay closer to the destination or
+            // broadcast to the rest of our section.
+            commands.extend(self.relay_message(&msg)?);
+        }
+        if !in_dst_location {
+            // Message not for us.
+            return Ok(commands);
+        }
+
         // Filter messages which were already handled
         if self.msg_filter.contains_incoming(&msg) {
             trace!("not handling message - already handled: {:?}", msg);
-            return Ok(None);
+            return Ok(commands);
         }
 
         match self.decide_message_status(&msg)? {
             MessageStatus::Useful => {
-                trace!("Useful message from {}: {:?}", sender, msg);
-                self.update_section_knowledge(&msg).await?;
-                self.handle_useful_message(Some(sender), msg).await
+                trace!("Useful message from {:?}: {:?}", sender, msg);
+                commands.extend(self.update_section_knowledge(&msg)?);
+                commands.extend(self.handle_useful_message(sender, msg)?);
```

Comment:
any diff to use `extend` here and `push` below?

Reply:
Because some function return `Vec<Command>`, some `Option<Command>` and some just `Command`.